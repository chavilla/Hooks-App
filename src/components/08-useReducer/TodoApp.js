import React, { useEffect, useReducer } from "react";
import FormTodo from "./FormTodo";
import "./styles.css";
import TodoList from "./TodoList";
import { todoReducer } from "./todoReducer";

const init = () => {
  return JSON.parse(localStorage.getItem("todos")) || [];
};

//Component
const TodoApp = () => {

  const [todos, dispatch] = useReducer(todoReducer, [], init);

  useEffect(() => {
    localStorage.setItem("todos", JSON.stringify(todos));
  }, [todos]);


  const handleAddTodo=(newTodo)=>{
    dispatch({
      type: "add",
      payload: newTodo,
    });
  }

  const handleToggle=(todoId)=>{
    dispatch({
      type:'toggle',
      payload:todoId
    })
  }

  const handleDelete = (id) => {
    dispatch({
      type: "delete",
      payload: id,
    });
  };

  return (
    <div>
      <h1>TodosApp {todos.length} </h1>
      <hr />
      <div className="row">
        <div className="col-7  d-flex flex-column justify-content-center">
          <h2 className="">Todos</h2>
          <ul className="d-flex flex-column justify-content-start p-0">
            {todos.map((todo, i) => (
              <TodoList
              key={todo.id }
              i={i}
              todo={todo}
              handleToggle={handleToggle}
              handleDelete={handleDelete}
              />
            ))}
          </ul>
        </div>
        <FormTodo
          handleAddTodo={handleAddTodo}
        />
      </div>
    </div>
  );
};

export default TodoApp;
