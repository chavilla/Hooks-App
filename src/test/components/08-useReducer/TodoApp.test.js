import { act } from "@testing-library/react";
import React from "react";
const { shallow, mount } = require("enzyme");
import TodoApp from "../../../components/08-useReducer/TodoApp";
import { state } from "../../fixtures/toDoList";

describe("<TodoApp />", () => {
  const wrapper = shallow(<TodoApp />);

  Storage.prototype.setItem=jest.fn(()=>{});

  test("should be displayed correctly", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("should add a toDo", () => {
    const wrapper = mount(<TodoApp />);

    act(() => {
      wrapper.find("FormTodo").prop("handleAddTodo")(state[0]);
      wrapper.find("FormTodo").prop("handleAddTodo")(state[1]);
    });

    expect(wrapper.find('h1').text()).toBe(`TodosApp ${state.length}`);
    expect(localStorage.setItem).toHaveBeenCalledTimes(2);

  });
});
