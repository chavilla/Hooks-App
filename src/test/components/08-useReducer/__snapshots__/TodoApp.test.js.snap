// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TodoApp /> should be displayed correctly 1`] = `
<div>
  <h1>
    TodosApp 
    0
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-7  d-flex flex-column justify-content-center"
    >
      <h2
        className=""
      >
        Todos
      </h2>
      <ul
        className="d-flex flex-column justify-content-start p-0"
      />
    </div>
    <FormTodo
      handleAddTodo={[Function]}
    />
  </div>
</div>
`;
